File: jsrc/legacy/SxxMachine/builtin/PRED_$cut_1.java
package SxxMachine;

import SxxMachine.IntegerTerm;
import SxxMachine.Operation;
import SxxMachine.Predicate;
import SxxMachine.Prolog;
import SxxMachine.Term;
import SxxMachine.IllegalTypeException;
/**
 <code>'$cut'/1</code><br>
 @author Mutsunori Banbara (banbara@kobe-u.ac.jp)
 @author Naoyuki Tamura (tamura@kobe-u.ac.jp)
 @version 1.0
*/
public class PRED_$cut_1 extends Predicate.P1 {
    public PRED_$cut_1(Term a1, Operation cont) {
        LARG[0] = a1;
        this.cont = cont;
    }

    public Operation exec(Prolog engine) {
	//        engine.setB0(); 
        Term a1;
        a1 = LARG[0];
        a1 = a1.dref();
        if (! (a1 instanceof IntegerTerm)) {
            throw new IllegalTypeException("integer", a1);
        } else {
            engine.cut(a1.asIntegerTerm().intValue());
        }
        return cont;
    }
}

File: src/legacy/SxxMachine/builtin/PRED_$cut_1.cpp
using namespace std;

#include "PRED_$cut_1.h"
#include "../../../machine/SxxMachine/Operation.h"
#include "../../../machine/SxxMachine/Term.h"
#include "../../../machine/SxxMachine/Prolog.h"
#include "../../../machine/SxxMachine/IntegerTerm.h"
#include "../../../exceptions/SxxMachine/IllegalTypeException.h"

namespace SxxMachine
{
	using IntegerTerm = SxxMachine::IntegerTerm;
	using Operation = SxxMachine::Operation;
	using Predicate = SxxMachine::Predicate;
	using Prolog = SxxMachine::Prolog;
	using Term = SxxMachine::Term;
	using IllegalTypeException = SxxMachine::IllegalTypeException;

	PRED_$cut_1::PRED_$cut_1(Term *a1, Operation cont)
	{
		LARG[0] = a1;
		this->cont = cont;
	}

	Operation PRED_$cut_1::exec(Prolog *engine)
	{
	//        engine.setB0(); 
		Term *a1;
		a1 = LARG[0];
		a1 = a1->dref();
		if (!(dynamic_cast<IntegerTerm*>(a1) != nullptr))
		{
			throw IllegalTypeException("integer", a1);
		}
		else
		{
			engine->cut(a1->asIntegerTerm()->intValue());
		}
		return cont;
	}
}
